lib: support dumb terminals

wlodzislav opened the pull request:
Add checks for `TERM='dumb'` in `console` and `repl`.

Add support for dumb terminals in `readline`. When `TERM='dumb'`:
- it doesn't use any ANSI escape codes for movement and output
- it doesn't use color
- it ignores all special keys except: `escape`, `return` and `ctrl-c`

##### Checklist
- [x] `make -j4 test` (UNIX), or `vcbuild test` (Windows) passes
- [x] tests  are included
- [x] commit message follows commit guidelines

BridgeAR commented on lib/readline.js:
```diff
@@ -768,6 +780,49 @@ Interface.prototype._moveCursor = function(dx) {
+
+    default:
+      if (s instanceof Buffer)
+        s = s.toString('utf-8');
```
This is currently never triggered and should likely be removed in the original function as well. This function should only be triggered with strings.

silverwind commented:
@BridgeAR do we need this? Maybe these cases can be covered via `hasColor` from https://github.com/nodejs/node/pull/26247.

BridgeAR commented on lib/readline.js:
```diff
@@ -768,6 +780,49 @@ Interface.prototype._moveCursor = function(dx) {
+      if (s instanceof Buffer)
+        s = s.toString('utf-8');
+
+      if (s) {
```
This if statement should be obsolete. Even if `s` is indeed an empty string (which would be surprising, the behavior would still stay the same).

wlodzislav replied:
It could, at least for `escape` key, also there is an example in https://nodejs.org/api/readline.html#readline_rl_write_data_key
That calls .write with first arg null.

```
rl.write(null, { ctrl: true, name: 'u' });
```

BridgeAR replied:
`null` would not end up in that place in the original function due to being detected as control sequence. That is not the case here but other truthy input is just as faulty and causes issues as well (e.g., `true`). Thus we could just verify that the input is indeed a string (`if (typeof s === 'string) {`) and ignore it otherwise.

wlodzislav commented:
It is different form `hasColor`. Dumb terminals doesn't support any ANSI escape codes(not only color codes).

According to `infocmp dumb` dumb terminals support only escape and return.

`readline` uses escape codes to redraw full line every keypress it it's run in tty, regardless of colors.

wlodzislav commented:
It is annoying when you run node repl or any cmd with console.log/dir inside GUI versions of VIM or emacs. It makes running node directly from them unusable.

BridgeAR commented on lib/readline.js:
```diff
@@ -181,7 +181,11 @@ function Interface(input, output, completer, terminal) {
+      self._ttyWriteDumb(s, key);
+    } else {
+      self._ttyWrite(s, key);
+    }
```
I personally would check the env while starting the readline instance and just set `instance._ttyWrite = _ttyWriteDumb.bind(instance)`. That way we have to check less often for the env and in less places.

However, there is no right or wrong so far and it's discussed in https://github.com/nodejs/node/issues/26213 if we want to adhere to a specific way to check for envs.

wlodzislav replied:
Ok, it will eliminate 2 checks, one in the frequent code path.

BridgeAR commented:
@silverwind as pointed out by @wlodzislav this is indeed required to really support `dumb` terminals. The `hasColors` function is still useful in this context (otherwise we'd have to change more places throughout the code base).

BridgeAR commented on lib/repl.js:
```diff
@@ -494,14 +494,17 @@ function REPLServer(prompt,
 
   if (options.useColors === undefined) {
-    options.useColors = self.terminal;
+    options.useColors = self.terminal && process.env.TERM !== 'dumb';
```
This could use the `writeStream.getColorDepth()` function to test for color support. It detects more terminals that lack color support and adds color support to dumb terminals that still support colors (`env.TERM = 'dumb' && env.COLORTERM = '1'`).

wlodzislav replied:
I think original problem is because programs use `dumb` to signal the apps preferred behavior, but don't clean any env variables. So it's used completely unrelated with colors.

For example in MacVim, when you run node with `:!` there are variables that may suggest color support:

```
  TERM_PROGRAM: 'iTerm.app',
  TERM: 'dumb',
  LSCOLORS: 'Gxfxcxdxbxegedabagacad',
  COLORS: '16777216',
  COLORFGBG: '7;0',
  COLORTERM: 'truecolor',
```

BridgeAR replied:
Yes, it seems like I made a mistake there. I'll open a PR soon to fix that.

BridgeAR replied:
https://github.com/nodejs/node/pull/26264

wlodzislav replied:
Replaced check by `.getColorDepth()` check similar to the one in the console.

wlodzislav commented on lib/repl.js:
```diff
@@ -494,14 +494,17 @@ function REPLServer(prompt,
-    self.writer.options = { ...writer.options, colors: true };
+    self.writer.options = {
+      ...writer.options,
+      colors: self.useColors
```
Ok. I'll change.

wlodzislav commented on lib/readline.js:
```diff
@@ -768,6 +780,49 @@ Interface.prototype._moveCursor = function(dx) {
   }
 };
 
+Interface.prototype._ttyWriteDumb = function(s, key) {
```
Ok. I'll combine it with `instance._ttyWrite = _ttyWriteDumb.bind(instance)` approach.

wlodzislav commented on lib/internal/console/constructor.js:
```diff
@@ -253,7 +253,7 @@ const kNoColorInspectOptions = {};
-    color = stream.isTTY && (
+    color = process.env.TERM !== 'dumb' && stream.isTTY && (
       typeof stream.getColorDepth === 'function' ?
         stream.getColorDepth() > 2 : true);
```
Not sure what you meant. Could you rephrase it, please?

In `getColorDepth` `env.COLORTERM` "overwrites" `env.TERM`:

```
  if (env.NODE_DISABLE_COLORS || env.TERM === 'dumb' && !env.COLORTERM) {
```

Maybe remove `&& !env.COLORTERM` part?

BridgeAR replied:
Refs: https://github.com/nodejs/node/pull/26261#discussion_r259415614

wlodzislav replied:
Removed dumb check.

silverwind commented:
> this is indeed required to really support dumb terminals

Then I would suggest something like `tty.supportsEscapes()` or similar and calling that I think its result could be statically determined at startup. Having separate write methods `_ttyWriteDumb` seems like nonsense.

wlodzislav commented:
> Having separate write methods `_ttyWriteDumb` seems like nonsense.

I originally thought to use `_normalWrite`, but it doesn't do the job - it ignores all incoming keys.

Dumb terminals couldn't process escape codes, but they could send any keys(like C-c).

And using `_ttyWrite` means that all branches will be wrapped in `if(dumb)` check to ignore all incoming keys that trigger commands that work with the cursor(basically all of them).

BridgeAR commented:
@silverwind 
> I would suggest adding something like `tty.supportsEscapes()`

But what does that function do internally? I guess it would just return `process.env.TERM !== 'dumb'`.

> Having separate write methods `_ttyWriteDumb` seems like nonsense.

We could theoretically just skip mostly everything and add more code to `_ttyWrite()` but that just increases the complexity of an already complex function. Separating the logic seems therefore right to me.

wlodzislav commented:
- Removed `_ttyWriteDumb` from prototype + replace `_ttyWrite` in constructor
- Removed Buffer check
- Added `s` string check + check if `s` not empty
- Removed unnecessary check for `useColors` in `repl`

Adding only `typeof` check makes `parallel/test-repl-editor.js` fail. When `_ttyWrite` is called with `s=''` there is an unnescessary call to `this._writeToOutput(s)` without non-empty check.

silverwind commented:
> But what does that function do internally? I guess it would just return process.env.TERM !== 'dumb'.

Oh, I was under the assumption that is ecompasses more than just that, but I guess if it is *that* simple, it might make sense to not move it to a method.

silverwind commented:
> Separating the logic seems therefore right to me.

On second thought, it probably makes sense, if we can avoid too much code duplication.

BridgeAR commented:
Ping @wlodzislav this needs a rebase and I landed my upstream fix to improve the color detection. So now it should be possible to address https://github.com/nodejs/node/pull/26261#discussion_r259375327 and https://github.com/nodejs/node/pull/26261#discussion_r259344759.

This is already in a great state and to me it seems it only needs a little bit more polishing.

BridgeAR commented:
@nodejs/tsc PTAL

wlodzislav commented on lib/internal/console/constructor.js:
```diff
@@ -366,7 +366,7 @@ Console.prototype.assert = function assert(expression, ...args) {
   // It only makes sense to clear if _stdout is a TTY.
   // Otherwise, do nothing.
-  if (this._stdout.isTTY) {
+  if (process.env.TERM !== 'dumb' && this._stdout.isTTY) {
```
Not sure about it. There are 2 cases when TERM="dumb" could apper:
- TERM is not set and parent program couldn't detect terminal features and set TERM="dumb"
- parent program intentionally set TERM="dumb" to limit output features

I guess the first case is rear and mostly it's the second one.
If we'll go with the second case as the relevant one, then warning isn't needed, because it's expected that the program(node) just will omit any escape codes in the output.

For the first case, there is already the detection code in `getColorDepth` and it silently defaults to COLORS_2 when can't detect the term.

BridgeAR replied:
It's probably not necessary. I just thought it might be weird if someone actively calls the function and nothing happens. But I am fine without it.

BridgeAR commented:
CI https://ci.nodejs.org/job/node-test-pull-request/21241/

BridgeAR commented:
@nodejs/tsc PTAL since this is semver-major. The PR is open for 14 days, marked as semver-major without a single TSC member having a look and I pinged the TSC before :/

Trott commented:
@BridgeAR Can you explain why this is semver-major? It seems to me that if it only changes behavior for dumb terminals by removing escape codes etc., that's a bug fix. On the other hand, I can certainly see marking it as semver-major out of caution. Is that it? Or is it semver-major for another reason that I'm failing to take note of?

silverwind commented on lib/readline.js:
```diff
@@ -1007,10 +1054,7 @@ Interface.prototype._ttyWrite = function(s, key) {
-          s = s.toString('utf-8');
-
-        if (s) {
+        if (typeof s === 'string' && s) {
```
Why this change? Is this code path never going to get buffers?

wlodzislav replied:
It could get `null` and empty strings.

wlodzislav replied:
`null` when triggering specials keys, see https://nodejs.org/api/readline.html#readline_rl_write_data_key

Empty strings are send when `readline` is used by `repl`, removing `&& s` check fails editor test for `repl`. Also it'll make it redraw the line unnecessarily.

BridgeAR replied:
@silverwind at least no core code would ever call this function with a buffer. It could theoretically be called directly by a user with a buffer. That would be surprising though and it would also be the only code path in the whole function that supports buffers. All others would not handle them and therefore it did not seem right to check for buffers (it was a legacy check from ~6-7 years ago).

silverwind replied:
Okay, sounds fine.

BridgeAR commented:
@Trott it is mainly the color check in the REPL that I would call semver-major. It checks if the stream supports colors or not. Before it just always expected color support. We could call that a bug fix but some people might rely on it.

The removed buffer check could also be counted towards being semver-major. It is very unlikely that anyone would indeed call this function from userland directly and pass in buffers but it is theoretically possible.

I am fine to handle this as semver-minor if the TSC agrees.

BridgeAR commented:
@jasnell me neither but it is a fixed expression and we have to use that terminology. See https://en.wikipedia.org/wiki/Computer_terminal#Dumb_terminals

Fishrock123 commented:
Is there a way to override color depth to ensure colors are outputted even if the terminal suggests it does not output color?

BridgeAR commented:
@Fishrock123 see https://github.com/nodejs/node/pull/26485.

This PR is somewhat independent of that though. A `dumb` terminal does not only not support colors, it is not capable of handling control sequences either.

BridgeAR commented:
@Trott @nodejs/tsc is the TSC fine with landing this as semver-minor instead of semver-major?

rvagg commented:
I'm not super keen on landing as a minor, because it's not. I don't feel strongly enough to argue it out but my default would be to stick with semver-major. A hypothetical situation is that someone is depending on parsing output in a particular way and they happen to have `dumb` set but don't know it, then suddenly the output is dramatically changed with an minor upgrade. I don't know the last time I saw `dumb`, though, but it's certainly plausable.

BridgeAR commented:
@rvagg in that case this PR only requires two LGs from the @nodejs/tsc.

BridgeAR commented:
@nodejs/tsc PTAL. This PR has a couple LGs but it misses one more TSC LG because it's semver-major.

Fishrock123 commented on lib/repl.js:
```diff
@@ -501,7 +501,9 @@ function REPLServer(prompt,
-    options.useColors = self.terminal;
+    options.useColors = self.terminal && (
+      typeof self.outputStream.getColorDepth === 'function' ?
+        self.outputStream.getColorDepth() > 2 : true);
```
Shouldn't this be `> 1`?

BridgeAR replied:
Good catch while it won't have any actual impact (the return values are either `1`, `4`, `8` or `24`).

wlodzislav replied:
I got that check from `lib/internal/console/constructor.js:256-258`, should I change the check there too in this PR?

addaleax replied:
@wlodzislav I think it’s okay to leave that for a separate PR, but here it would be good to replace it with `> 1` before this is being merged.

BridgeAR replied:
I fixed this while landing.

BridgeAR commented:
Landed in 5f032a7a269b66d48505869b0ae4fb1db403b118...82b3ee776ba87481b5fa3d00cbe7f3c4f32c3ec3 :tada:

@wlodzislav congratulations on your first commits to Node.js! Great work!
